DESIGN.md

```diff
@@ -0,0 +1,247 @@
+* Making sure the library will call it as a pass-thru/wrapper command for the user-provided command
+  * `rockholla-go-linux-ns-cgroup-worker-fork -- ping -c 2 google.com` as an example of the command the library will call via `exec.Command`
+* This fork/wrapper will do the necessary per-process-level setup via the compiled Golang code of namespace isolation/cgroup limits for our needs before calling the user-provided command to be run w/in the isolated namespace/cgroup:
+  * `cgroup` setup at `/sys/fs/cgroup/**/*` for our resource control needs of a single process
+  * `CLONE_NEWNS`: we will need to set up a dedicated root filesystem for use by the process, and [`PivotRoot` to it](https://pkg.go.dev/golang.org/x/sys/unix#PivotRoot) (`pivot_root` system call), and set up some expected paths to common system/user binary paths, e.g. `/usr/bin` `/usr/local/bin` etc.
+  * `CLONE_NEWPID`: we will need to mount a dedicated `/proc` for isolated process/PID awareness
+  * `CLONE_NEWNET`: we will need to set up network interfaces dedicated to the namespace
+  * Assigning the forked process ID to `/sys/fs/cgroup/pids/[users-process-id]/cgroup.procs`
+* Our library will call this fork/wrapper binary with the pass-thru of the user command via `os/exec.Command` to fork and wire the streamed output via our `ProcessStore` implementation `Stdout` and `Stderr` `io.Writer` implementations to `os.exec.Command.Stdout` and `os.exec.Command.Stderr`
+* This call to the fork will happen async, in a go routine, to ensure we can run it in the background, using a channel to communicate as necessary to the parent process in the library/the API process calling the library.
+
+### How Stopping an Existing Process Works
+
+The input to `Stop` is a service-specific worker ID. The library will take the following path to stop the related process if it can:
+
+* The `ProcessStore` and result of `Start` will be a service-specific worker ID
+* The `Stop` command will accept this service-specific worker ID, validate that the `requesterID` is allowed to stop the process, in our case only if its exactly the owner, retrieve the related host PID via the `ProcessStore.GetHostPID` implementation
+* Then an `exec.Command` to `kill` will be launched for that PID
+* Additionally we will perform cleanup for cgroup paths, `ProcessStore` entry, etc. and all other things that might not be captured by normal Golang garbage collection. This will be part of a resuable method that is utilized on completion/termination of a worker otherwise as well. 
+
+## API
+
+The API will be a gRPC-based API like the following proto definition for the service and request messages:
+
+```
+service Worker {
+  rpc Start (StartRequest) returns (StartResponse) {}
+  rpc Stop (StopRequest) returns (StopResponse) {}
+  rpc GetStatus (GetStatusRequest) returns (GetStatusResponse) {}
+  rpc StreamOutput(StreamOutputRequest) returns (stream StreamOutputResponse) {}
+}
+
+message StartRequest {
+  string cmd = 1;
```

Comment:
Does cmd also contain the command arguments? Does it make sense to split into cmd and arguments fields ?

Reply:
https://github.com/rockholla/go-linux-ns-cgroup-worker/pull/1/commits/28c29685d10a807b3e8eb9e0304aad38e0eaaebc